sky/packages/sky/lib/src/widgets/navigator2.dart

```diff
@@ -2,55 +2,56 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:async';
-
-import 'package:flutter/animation.dart';
-
-import 'basic.dart';
 import 'framework.dart';
 import 'overlay.dart';
-import 'transitions.dart';
 
 abstract class Route {
-  /// Override this function to return the widget that this route should display.
-  Widget createWidget();
+  List<Widget> createWidgets() => const <Widget>[];
+
+  OverlayEntry get topEntry => _entries.isNotEmpty ? _entries.last : null;
+  OverlayEntry get bottomEntry => _entries.isNotEmpty ? _entries.first : null;
 
-  Widget _child;
-  OverlayEntry _entry;
+  final List<OverlayEntry> _entries = new List<OverlayEntry>();
 
-  void willPush() {
-    _child = createWidget();
+  void add(OverlayState overlay, OverlayEntry insertionPoint) {
+    List<Widget> widgets = createWidgets();
+    for (Widget widget in widgets) {
+      _entries.add(new OverlayEntry(child: widget));
+      overlay?.insert(_entries.last, above: insertionPoint);
+      insertionPoint = _entries.last;
+    }
   }
 
-  void didPop(dynamic result) {
-    _entry.remove();
+  void didMakeCurrent() { }
+
+  void remove(dynamic result) {
+    for (OverlayEntry entry in _entries)
+      entry.remove();
   }
 }
 
-typedef Widget RouteBuilder(args);
-typedef RouteBuilder RouteGenerator(String name);
+class RouteArguments {
```

Comment:
Maybe NamedRouteSettings? The reason I asked about pushState() is that I was wondering if it would even bother to new up one of these RouteArguments objects. I'm guessing not, in which case RouteArguments isn't as generic as the name implies.